Delete messages written as a channel

hoosnick opened the pull request:
# Stop bloggers !

hoosnick committed:
added new variable

Added "you_are_blogger" variable with value "Kanal nomidan yozish mumkin emas.".

hoosnick committed:
Added new feature

Delete and alert bloggers who write as channel

hoosnick committed:
Update decorators.py

hoosnick committed:
New feature: Delete messages written as a channel

hoosnick committed:
Updated: fixed some bugs

JasurbekNURBOYEV commented on core/factory.py:
```diff
@@ -124,15 +124,35 @@ def is_member(self, uid: int, whitelist: List[int]) -> bool:
+        Checks if the user is writing as a channel
+        """
+        user = message.from_user
+        if user.first_name == 'Channel' and user.is_bot:
```
Since the only thing is being done is evaluation of this logical operation, we can replace the whole thing with a single line:
```python
return user.first_name == 'Channel' and user.is_bot
```

JasurbekNURBOYEV commented on core/factory.py:
```diff
@@ -124,15 +124,35 @@ def is_member(self, uid: int, whitelist: List[int]) -> bool:
+            return True
+        return False
+    
+    def notify_about_membership(self, message, is_blogger: bool = False) -> None:
```
Why are we modifying the existing method? That method has its own purpose. We should make a new one, something like "notify_about_violation".

JasurbekNURBOYEV commented on core/factory.py:
```diff
@@ -124,15 +124,35 @@ def is_member(self, uid: int, whitelist: List[int]) -> bool:
         finally:
             # if user is not a member, using the bot is considered violation
-            self.set_context(message, 'violation', True)
+            None if is_blogger else self.set_context(message, 'violation', True)
```
This is an expression, since we are not using its value, it is better not to use an expression here at all. This would be a better and more readable option:
```python
if is_blogger:
  self.set_context(message, 'violation', True)
```

JasurbekNURBOYEV commented on core/factory.py:
```diff
@@ -376,7 +396,7 @@ def is_commander(self, cid, uid) -> bool:
 # command handlers
 @bot.message_handler(commands=[constants.COMMAND_START, constants.COMMAND_CANCEL, constants.COMMAND_TEST])
-@lock_method_for_strangers(checker=bot.is_member, default=bot.notify_about_membership)
+@lock_method_for_strangers(checker=bot.is_member, stop_bloggers=bot.is_blogger, default=bot.notify_about_membership)
```
We shouldn't be modifying the existing decorator. Instead, we should re-apply it with a separate checker. We want something like this:
```python
@lock_method_for_strangers(checker=bot.is_member, default=bot.notify_about_membership)
@lock_method_for_strangers(checker=bot.is_blogger, default=bot.notify_about_violation)
def command_handler(message):
```

JasurbekNURBOYEV commented on core/factory.py:
```diff
@@ -592,7 +612,7 @@ def new_chat_member_handler(message):
 # here we try to handle text messages
 @bot.message_handler(content_types=['text'])
-@lock_method_for_strangers(checker=bot.is_member, default=bot.notify_about_membership)
+@lock_method_for_strangers(checker=bot.is_member, stop_bloggers=bot.is_blogger, default=bot.notify_about_membership)
```
Same issue as the previous one

JasurbekNURBOYEV commented on core/factory.py:
```diff
@@ -916,7 +936,7 @@ def text_handler(message):
 # photo handler
 @bot.message_handler(content_types=['photo'])
-@lock_method_for_strangers(bot.is_member, bot.notify_about_membership)
+@lock_method_for_strangers(bot.is_member, stop_bloggers=bot.is_blogger, bot.notify_about_membership)
```
Same here too (check the previous comment)

JasurbekNURBOYEV commented on utils/decorators.py:
```diff
@@ -12,12 +12,14 @@
 
 def lock_method_for_strangers(
         checker: Callable[[int, List[int]], bool],
+        stop_bloggers: Callable = None,
```
Please, do not change this decorator. As I mentioned above, we should keep this original decorator and re-apply it with a different checker method instead. You can revert all changes in this file.

JasurbekNURBOYEV reviewed (changes requested):
Nice job taking care of this task!
There are several places where we can improve the overall code structure and its readability. Please address the comments.

hoosnick committed:
Updated and replaced is_blogger() func

Updated and replaced return for is_blogger checker function proposed by Jasurbek

hoosnick committed:
Created new method "notify_about_violation"

hoosnick committed:
Update decorators.py

hoosnick committed:
Update factory.py

JasurbekNURBOYEV commented on core/factory.py:
```diff
@@ -124,7 +125,14 @@ def is_member(self, uid: int, whitelist: List[int]) -> bool:
+        Checks if the user is writing as a channel
+        """
+        user = message.from_user
+        return False if user.first_name == 'Channel' and user.is_bot else True
```
I think it is vice versa: this expression yields opposite result. This should be correct:
```python
return user.first_name == 'Channel' and user.is_bot
```

Also, we should be using `whitelist`. There seems to be `sender_chat` field of Message object, it should have `id` attribute, we can use that to check against. Something like this:
```python
if hasattr(message, "sender_chat") and message.sender_chat.id in whitelist:
  return False
return user.first_name == 'Channel' and user.is_bot
```

JasurbekNURBOYEV commented on core/factory.py:
```diff
@@ -133,7 +141,18 @@ def notify_about_membership(self, message) -> None:
+            # delete sent message as a channel
+            self.delete_message(message.chat.id, message.message_id)
+            # notify user about this
+            self.send_message(message.from_user.id, self.strings.you_are_blogger)
```
It seems `from_user.id` points to a bot called Channel, we can't send message to it. I don't think we can send message at all in this case. Let's remove that part only delete the message for now.

JasurbekNURBOYEV commented on core/strings.py:
```diff
@@ -11,6 +11,7 @@ class Strings(object):
     # strings related to testing process
     welcome = "Hey! Ahvollar qalay!"
     you_are_not_a_member = "Botdan foydalanish uchun asosiy guruh a'zosi bo'lishingiz kerak."
+    you_are_blogger = "Kanal nomidan yozish mumkin emas."
```
Since we are not sending message in that scenario, I think we can remove this string
